### DISCUSSION TITLE
System.Text.Json: How to get already mapped members when deserializing?

### DISCUSSION THREAD

**Discussion by albyho** [dotnet/runtime#96241] [General]
``` C#
public class Song {
  public int Id {get; set;}
  public string? Title {get; set;}
  public string? Foo {get; set;}
  public string? Bar {get; set;}
}

var song = JsonSerializer.Deserialize<Song>("""{"id": 1, "title": "Test", "foo": "a"}""");
```

How to know `song` contains only `id`, `title` and `foo`, not `bar`?

**Comment by ghost** [comment]
Tagging subscribers to this area: @dotnet/area-system-text-json, @gregsdennis
See info in [area-owners.md](https://github.com/dotnet/runtime/blob/main/docs/area-owners.md) if you want to be subscribed.
<details>
<summary>Issue Details</summary>
<hr />

``` C#
public class Song {
  public int Id {get; set;}
  public string? Title {get; set;}
  public string? Foo {get; set;}
  public string? Bar {get; set;}
}

var song = JsonSerializer.Deserialize<Song>("""{"id": 1, "title": "Test", "foo": "a"}""");
```

How to know `song` contains only `id`, `title` and `foo`, not `bar`?

<table>
  <tr>
    <th align="left">Author:</th>
    <td>albyho</td>
  </tr>
  <tr>
    <th align="left">Assignees:</th>
    <td>-</td>
  </tr>
  <tr>
    <th align="left">Labels:</th>
    <td>

`area-System.Text.Json`

</td>
  </tr>
  <tr>
    <th align="left">Milestone:</th>
    <td>-</td>
  </tr>
</table>
</details>

**Comment by gregsdennis** [comment]
I thhink you have a few options.

### Just deserialize it

Deserializing as you have in your code sample will result in `song.Bar` being null.  The downside here is that you can't tell the difference between `bar` missing from the JSON or `bar` being present with a JSON-null value.

### Parse first

You can parse the JSON text into `JsonNode` or `JsonDocument` first.  These are JSON DOM types that allow you to examine the structure of the JSON without deserializing it.  Then you can simply check the DOM to see if a `bar` property exists.

I find `JsonNode` easier to work with, but `JsonDocument` will give better performance because it's not allocating anything.

### JSON Schema

This builds on the previous option since you still have to parse the JSON.

Use something like [_JsonSchema.Net_](https://docs.json-everything.net/schema/basics/) (disclaimer: it's mine) to define constraints that you want for the JSON with JSON Schema.  Then you can evaluate the JSON with the schema, and you'll get output that indicates whether the JSON meets the required constraints.

**Comment by eiriktsarpalis** [comment]
> Deserializing as you have in your code sample will result in `song.Bar` being null. The downside here is that you can't tell the difference between `bar` missing from the JSON or `bar` being present with a JSON-null value.

That's probably what I would do as well. There shouldn't be a compelling semantic difference between `bar` being explicitly set to `null` in the JSON or missing since it's not representible in the model but FWIW you can exclude the latter possibility by marking the property as `required` (or using the `JsonRequiredAttribute`).

**Comment by TonyValenti** [comment]
I have a custom type I've created called optional<T> which lets me distinguish between null and not-set values.
